transaction.go

```diff
@@ -0,0 +1,158 @@
+package seth
+
+import (
+	"bytes"
+	"math/big"
+)
+
+type Encoding int
+
+const (
+	LIST Encoding = iota
+	STRING
+	INTEGER
+)
+
+type SignedTransaction struct {
+	Tx  *Transaction
+	Sig *Signature
+}
+
+func intBytes(n Uint64) []byte {
+	var buf bytes.Buffer
+	shift := uint64(0)
+
+	for (n >> shift) > 0xFF {
+		shift += 8
+	}
+
+	for shift > 0 {
+		buf.Write([]byte{byte(n >> shift)})
+		shift -= 8
+	}
+	buf.Write([]byte{byte(n)})
+
+	return buf.Bytes()
+}
+
+func Encode(b interface{}, e Encoding) []byte {
+	switch e {
+	case LIST:
+		val := b.([]byte)
+		if len(val) == 1 && val[0] == 0x00 {
+			return []byte{0x80}
+		}
+		return encodeBytes(val, 0xC0, 0xF7)
+	case STRING:
+		val := b.([]byte)
+		if bytes.Equal(val, []byte("")) {
+			return []byte{0x80}
+		} else if bytes.Equal(val, []byte{0x00}) {
+			return []byte{0x00}
+		}
+		return encodeBytes(b.([]byte), 0x80, 0xB7)
+	case INTEGER:
+		n := b.(Uint64)
+		if n == 0 {
+			return []byte{0x80}
+		}
+		return encodeBytes(intBytes(n), 0x80, 0xB7)
+	}
+	return []byte{}
+}
+
+func encodeBytes(b []byte, sh byte, lh byte) []byte {
+	var buf bytes.Buffer
+
+	if len(b) == 1 && b[0] < 0x80 {
+		return b
+	}
+
+	if len(b) < 56 {
+		buf.Write([]byte{byte(int(sh) + len(b))})
+	} else {
+		blen := intBytes(Uint64(len(b)))
+		buf.Write([]byte{byte(int(lh) + len(blen))})
+		buf.Write(blen)
+	}
+
+	buf.Write(b)
+	return buf.Bytes()
+}
+
+func (t *Transaction) rawTx() ([]byte, error) {
+	var data bytes.Buffer
+
+	var toAddr []byte
+	if t.To == nil {
+		toAddr = []byte("")
+	} else {
+		toAddr = t.To[:]
+	}
+
+	data.Write(Encode(t.Nonce, INTEGER))
+	data.Write(Encode(t.GasPrice.Big().Bytes(), STRING))
+	data.Write(Encode(t.Gas, INTEGER))
+	data.Write(Encode(toAddr, STRING))
+	data.Write(Encode(t.Value.Big().Bytes(), STRING))
+	data.Write(Encode([]byte(t.Input), STRING))
+
+	return data.Bytes(), nil
+}
+
+func (t *Transaction) MarshalBinary() ([]byte, error) {
+	var res bytes.Buffer
+
+	data, err := t.rawTx()
+	if err != nil {
+		return nil, err
+	}
+
+	res.Write(Encode(data, LIST))
+	return res.Bytes(), nil
+}
+
+func (t *SignedTransaction) MarshalBinary() ([]byte, error) {
+	var res bytes.Buffer
+	var buf bytes.Buffer
+
+	data, err := t.Tx.rawTx()
+	if err != nil {
+		return nil, err
+	}
+
+	var r, s big.Int
+	r.SetBytes(t.Sig[:32])
+	s.SetBytes(t.Sig[32:64])
+
+	buf.Write(data)
+	buf.Write(Encode(Uint64(t.Sig[64]), INTEGER))
+	buf.Write(Encode(r.Bytes(), STRING))
+	buf.Write(Encode(s.Bytes(), STRING))
+
+	res.Write(Encode(buf.Bytes(), LIST))
+	return res.Bytes(), nil
+}
+
+func (t *Transaction) Sign(p *PrivateKey) (*Hash, *Signature, error) {
```

Comment:
I think this should be re-phrased as
```go
func SignTransaction(tx *Transaction, s crypto.Signer) ([]byte, error)
```
since the `crypto` library in the standard library has a generic signer interface, and since we anticipate adding an HSM implementation of `crypto.Signer` in the near future.

With the existing code, you'd use that function like
```
key := seth.NewPrivateKey()
out, err := seth.SignTransaction(tx, key.ToECDSA())
```
I think we can do away with the `SignedTransaction` type as well, and just expose this API as a way from going from a `seth.Transaction` structure to an (opaque) signed transaction. You only need the raw binary to call `eth_sendRawTransaction` anyway, so there's not much point in having an intermediate representation.